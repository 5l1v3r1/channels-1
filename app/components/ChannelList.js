// @flow
import React from 'react';
import BlankState from './core/BlankState';
import Channel from './core/Channel';
import ErrorState from './core/ErrorState';
import Loader from './core/Loader';
import type { PromiseAction } from '../constants/types';
import ChannelType from '../dataTypes/channelType';
import styles from './ChannelList.scss';

const ChannelList = ({
  channels,
  activeChannelId,
  isFetching,
  error,
  channelListIsOpen,
  fetchChannels,
  updateActiveChannel,
  toggleChannelList
}: {
  channels: ChannelType[],
  activeChannelId: string,
  isFetching: boolean,
  error: boolean,
  channelListIsOpen: boolean,
  fetchChannels: () => PromiseAction,
  updateActiveChannel: (channelId: string) => void,
  toggleChannelList: () => void
}) => (
  <div className={channelListIsOpen ? [styles.channelList, styles.isOpen].join(' ') : styles.channelList}>
    <div className={styles.closeTarget} onClick={toggleChannelList} />
    <div className={styles.modal}>
      {isFetching || error ? (
        isFetching ? (
          <Loader />
        ) : (
          <ErrorState
            color="black"
            message="Error requesting channels."
            retry={fetchChannels}
          />
        )
      ) : (
        channels.length ? (
          <ul className={styles.channels} onClick={toggleChannelList}>
            {channels.map(channel => (
              <Channel
                key={channel.id}
                title={channel.title}
                thumbnail={channel.thumbnail}
                videoCount={channel.videoCount}
                subscriberCount={channel.subscriberCount}
                isActive={channel.id === activeChannelId}
                handleClick={() => updateActiveChannel(channel.id)}
              />
            ))}
          </ul>
        ) : (
          <BlankState
            color="black"
            message="No channels found."
          />
        )
      )}
    </div>
  </div>
);

export default ChannelList;
