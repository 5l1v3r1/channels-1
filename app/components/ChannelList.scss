@import '../app.global.scss';

.channelListIsOpen {
	.closeTarget {
		position: fixed;
		top: 0;
		left: 0;
		width: 100%;
		height: 100%;
	}

	.modal {
		opacity: 1;
		visibility: visible;
		transition: visibility 0s, opacity 0.3s ease;
	}
}

.modal {
  position: fixed;
  top: 75px;
  left: 11px;
  width: 350px;
  height: 450px;
  border-radius: 3px;
  box-shadow: 0 2px 15px rgba(46, 46, 46, 0.6);
  background-color: $white;
  overflow-y: auto;
  opacity: 0;
  visibility: hidden;
}

.channels {
  margin: 0;
  padding: 0;
  list-style: none;
}
