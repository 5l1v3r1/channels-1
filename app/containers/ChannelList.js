import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as channelsActions from '../actions/channels';
import ChannelList from '../components/ChannelList';
import { getChannels } from '../reducers/index';

function mapStateToProps(state, ownProps) {
  return {
    channels: getChannels(state),
    isFetching: state.channels.isFetching,
    error: state.channels.error,
    activeChannelId: ownProps.activeChannelId,
    channelListIsOpen: ownProps.channelListIsOpen,
    toggleChannelList: ownProps.toggleChannelList
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(channelsActions, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(ChannelList);
