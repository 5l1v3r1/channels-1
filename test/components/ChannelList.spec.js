import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ChannelList from '../../app/components/ChannelList';

function setup() {
  const props = {
    channels: [],
    activeChannelId: '',
    isFetching: false,
    error: false,
    channelListIsOpen: false,
    fetchChannels: jest.fn(),
    updateActiveChannel: jest.fn(),
    toggleChannelList: jest.fn()
  };

  const enzymeWrapper = shallow(<ChannelList {...props} />);

  return {
    props,
    enzymeWrapper
  };
}

describe('ChannelList', () => {
  describe('when closed', () => {
    it('should render self', () => {
      const { enzymeWrapper } = setup();
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when opened', () => {

    const { props, enzymeWrapper } = setup();
    beforeEach(() => {
      enzymeWrapper.setProps({
        channelListIsOpen: true
      });
    });

    afterEach(() => {
      enzymeWrapper.setProps({
        ...props
      });
    });

    it('should render self with loader', () => {
      enzymeWrapper.setProps({
        isFetching: true
      });
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });

    it('should render self with error message', () => {
      enzymeWrapper.setProps({
        error: true
      });
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });

    it('should render self with channels', () => {
      enzymeWrapper.setProps({
        channels: [
          {
            id: 'UCyIe-61Y8C4_o-zZCtO4ETQ',
            subscriberCount: '243,643',
            thumbnail: 'https://yt3.ggpht.com/photo.jpg',
            title: 'DevTips',
            videoCount: '292'
          },
          {
            id: 'UC7O6CntQoAI-wYyJxYiqNUg',
            subscriberCount: '18,655',
            thumbnail: 'https://yt3.ggpht.com/photo.jpg',
            title: 'Ihatetomatoes',
            videoCount: '140'
          }
        ]
      });
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });

    it('should render self with no channels', () => {
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });
  });

  it('should call toggleChannelList', () => {
    const { props, enzymeWrapper } = setup();
    enzymeWrapper.setProps({
      channels: [{
        id: 'UCyIe-61Y8C4_o-zZCtO4ETQ',
        subscriberCount: '243,643',
        thumbnail: 'https://yt3.ggpht.com/photo.jpg',
        title: 'DevTips',
        videoCount: '292'
      }]
    });
    const closeTarget = enzymeWrapper.find('.closeTarget');
    const channels = enzymeWrapper.find('.channels');

    closeTarget.simulate('click');

    expect(props.toggleChannelList).toHaveBeenCalledTimes(1);

    channels.simulate('click');

    expect(props.toggleChannelList).toHaveBeenCalledTimes(2);
  });
});
