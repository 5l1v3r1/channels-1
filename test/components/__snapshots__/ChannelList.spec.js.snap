// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelList should render self with channels 1`] = `
<div
  className="channelList isOpen"
>
  <div
    className="closeTarget"
    onClick={[Function]}
  />
  <div
    className="modal"
  >
    <ul
      className="channels"
      onClick={[Function]}
    >
      <Channel
        handleClick={[Function]}
        isActive={false}
        key="1"
        subscriberCount="1"
        thumbnail="thumbnail.jpg"
        title="title"
        videoCount="1"
      />
      <Channel
        handleClick={[Function]}
        isActive={false}
        key="2"
        subscriberCount="1"
        thumbnail="thumbnail.jpg"
        title="title"
        videoCount="1"
      />
    </ul>
  </div>
</div>
`;

exports[`ChannelList should render self with error 1`] = `
<div
  className="channelList isOpen"
>
  <div
    className="closeTarget"
    onClick={[Function]}
  />
  <div
    className="modal"
  >
    <ErrorState
      color="black"
      message="Error requesting channels."
      retry={[Function]}
    />
  </div>
</div>
`;

exports[`ChannelList should render self with loader 1`] = `
<div
  className="channelList isOpen"
>
  <div
    className="closeTarget"
    onClick={[Function]}
  />
  <div
    className="modal"
  >
    <Loader />
  </div>
</div>
`;

exports[`ChannelList should render self with no channels 1`] = `
<div
  className="channelList isOpen"
>
  <div
    className="closeTarget"
    onClick={[Function]}
  />
  <div
    className="modal"
  >
    <BlankState
      color="black"
      message="No channels found."
    />
  </div>
</div>
`;
